### PR TITLE
validate fromversion, toversion in content items

### PR DESCRIPTION
**Pull request by israelpoli**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/48225

## Description
- Added a validation that checks if the 'fromversion' or 'toversion' key in a content item are incorrect format.
- Added a validation that checks if `fromversion` <= `toversion`.

## Must have
- [x] Tests

### CODE REVIEW

File: demisto_sdk/commands/common/errors.py [hunk 1]
```diff
@@ -2082,6 +2083,12 @@ def taskid_different_from_id(task_key, id_, taskid):
         return f"On task: {task_key},  the field 'taskid': {taskid} and the 'id' under the 'task' field: {id_}, " \
                f"must be with equal value. "
 
+    @staticmethod
+    @error_code_decorator
+    def from_and_to_version_are_incorrect_format(incorrect_key):
```
Comment (dorschw) [inline demisto_sdk/commands/common/errors.py:2088]:
```suggestion
    def from_and_to_version_are_incorrect_format(incorrect_key: str):
```

File: demisto_sdk/commands/common/hook_validations/test_playbook.py [hunk 2]
```diff
@@ -24,6 +27,7 @@ def is_valid_test_playbook(self, validate_rn: bool = False) -> bool:
             self.is_valid_file(validate_rn),
             self._is_id_uuid(),
             self._is_taskid_equals_id(),
+            self.are_fromversion_and_toversion_in_correct_format()
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/test_playbook.py:30]: how do we validate it in other content items?

File: demisto_sdk/commands/common/hook_validations/test_playbook.py [hunk 3]
```diff
@@ -35,6 +39,22 @@ def is_valid_file(self, validate_rn):
             self.is_valid_fromversion(),
         ])
 
+    def are_fromversion_and_toversion_in_correct_format(self) -> bool:
+
+        version_structure = re.compile(FROM_TO_VERSION_REGEX)
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/test_playbook.py:44]: better compile it at the beginning of the file, or in constants

**Comment by idovandijk** [comment]
@dorschw do we want to have those fields in the test playbooks or in `conf.json`? Seems like we have a combination of both right now (one should be redundant).

**Comment by dorschw** [comment]
> @dorschw do we want to have those fields in the test playbooks or in `conf.json`? Seems like we have a combination of both right now (one should be redundant).

We do, but perhaps it's for a different issue. Please open one and we'll discuss it there. @israelpoli is moving the validation to be applied for all content items, so this PR isn't only about TPBs anyway.

**Comment by idovandijk** [comment]
> > @dorschw do we want to have those fields in the test playbooks or in `conf.json`? Seems like we have a combination of both right now (one should be redundant).
> 
> We do, but perhaps it's for a different issue. Please open one and we'll discuss it there. @israelpoli is moving the validation to be applied for all content items, so this PR isn't only about TPBs anyway.

ok
https://github.com/demisto/etc/issues/48260

File: demisto_sdk/commands/common/hook_validations/test_playbook.py [hunk 4]
```diff
@@ -35,6 +39,22 @@ def is_valid_file(self, validate_rn):
             self.is_valid_fromversion(),
         ])
 
+    def are_fromversion_and_toversion_in_correct_format(self) -> bool:
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/test_playbook.py:42]:
```suggestion
    @error_codes('PB118')
    def are_fromversion_and_toversion_in_correct_format(self) -> bool:
```

File: demisto_sdk/commands/common/hook_validations/content_entity_validator.py [hunk 5]
```diff
@@ -250,6 +254,52 @@ def should_run_fromversion_validation(self):
...
+            return True
+
+        if not FROM_TO_VERSION_REGEX.fullmatch(from_version):
+            error_message, error_code = Errors.from_and_to_version_are_incorrect_format(
+                'fromversion')
+            self.handle_error(error_message, error_code, file_path=self.file_path)
+            return False
+        if not FROM_TO_VERSION_REGEX.fullmatch(to_version):
+            error_message, error_code = Errors.from_and_to_version_are_incorrect_format(
+                'tovesion')
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/content_entity_validator.py:276]: typo - was it not caught in the UT?

File: demisto_sdk/commands/common/hook_validations/content_entity_validator.py [hunk 6]
```diff
@@ -250,6 +254,52 @@ def should_run_fromversion_validation(self):
...
+            self.handle_error(error_message, error_code, file_path=self.file_path)
+            return False
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/content_entity_validator.py:278]:
```suggestion
        for field, name in ((from_version,'fromversion'),(to_version,'toversion')):
            if not FROM_TO_VERSION_REGEX.fullmatch(field):
                error_message, error_code = Errors.from_and_to_version_are_incorrect_format(name)
                self.handle_error(error_message, error_code, file_path=self.file_path)
                return False
```

File: demisto_sdk/commands/common/hook_validations/content_entity_validator.py [hunk 7]
```diff
@@ -250,6 +254,52 @@ def should_run_fromversion_validation(self):
...
+    def are_fromversion_toversion_synchronized(self) -> bool:
+
+        if self.file_path.endswith('.json'):
+            from_version = self.current_file.get('fromVersion', '')
+            to_version = self.current_file.get('toVersion', '')
+        elif self.file_path.endswith('.yml'):
+            from_version = self.current_file.get('fromversion', '')
+            to_version = self.current_file.get('toversion', '')
+        else:
+            return True
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/content_entity_validator.py:292]: please add a debug logging here: `are_fromversion_toversion_synchronized run on a non json/yml file - returning True as default`

File: demisto_sdk/commands/common/hook_validations/content_entity_validator.py [hunk 8]
```diff
@@ -250,6 +254,52 @@ def should_run_fromversion_validation(self):
...
+
+        if not from_version or not to_version:
+            return True
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/content_entity_validator.py:295]: please add debug logging here too: `either not from_version or not to_version, considering them synced`

File: demisto_sdk/commands/common/tests/content_entity_validator_test.py [hunk 9]
```diff
@@ -229,3 +229,77 @@ def test_validate_readme_exists_not_checking_on_api_modules(repo):
...
+    (
+        {'toversion': '0.f.0'},
+        'test.yml',
+        False
+    ),
+    (
+        {'toversion': 'test'},
+        'test',
+        True
+    ),
```
Comment (dorschw) [inline demisto_sdk/commands/common/tests/content_entity_validator_test.py:284]: please add a test case of from/to version that are `''`

**Comment by idovandijk** [comment]
from comparison of fromversion to toversion did we check if it's works correctly with 3 decimals vs 2?
example: comparison from 5.0 to 5.5.0

**Comment by israelpoli** [comment]
@idovandijk
It also works on the example you have,
But it will never reach such a state following the second validation that validates the version format (xx.xx.xx)

**Review by dorschw — APPROVED** [review]

**Pull request merged**